proxy.go

```diff
@@ -193,6 +242,229 @@ func (proxy *proxy) releaseToken(token Token) (*tokenInfo, error) {
 	return info, nil
 }
 
+// returns false if it's a clean start (i.e. no state is stored) or restoring failed
+func (proxy *proxy) restoreState() bool {
+	proxyStateFilePath := storeStateDir + "proxy_state.json"
+	if _, err := os.Stat(storeStateDir); os.IsNotExist(err) {
+		err := os.MkdirAll(storeStateDir, proxyStateDirPerm)
+		if err != nil {
+			proxyLog.Errorf("Couldn't create directory %s: %v",
+				storeStateDir, err)
+		}
+		return false
+	}
+
+	fdata, err := ioutil.ReadFile(proxyStateFilePath)
+	if err != nil {
+		proxyLog.Errorf("Couldn't recover from %s: %v", proxyStateFilePath, err)
+		return false
+	}
+
+	var proxyState proxyStateOnDisk
+	err = json.Unmarshal(fdata, &proxyState)
+	if err != nil {
+		proxyLog.Errorf("Couldn't unmarshal %s: %v", proxyStateFilePath, err)
+		return false
+	}
+	proxyLog.Debugf("proxy: %+v", proxyState)
+
+	if len(proxyState.ContainerIDs) == 0 {
+		return false
+	}
+	proxyLog.Warn("Recovering proxy state from: ", proxyStateFilePath)
+	if proxyState.Version != Version {
+		proxyLog.Warnf("Stored state version (%s) mismatches proxy"+
+			" version (%s). Aborting", proxyState.Version, Version)
+		return false
+	}
+
+	proxy.socketPath = proxyState.SocketPath
+	proxy.enableVMConsole = proxyState.EnableVMConsole
+
+	for _, contID := range proxyState.ContainerIDs {
+		go restoreVMState(proxy, contID)
+	}
+
+	return true
+}
+
+func (proxy *proxy) storeState() {
+	proxyStateFilePath := storeStateDir + "proxy_state.json"
+	proxy.Lock()
+	defer proxy.Unlock()
+
+	// if there are 0 VMs then remove state from disk
+	if (len(proxy.vms)) == 0 {
+		if err := os.Remove(proxyStateFilePath); err != nil {
+			proxyLog.Errorf("Couldn't remove %s: %v",
+				proxyStateFilePath, err)
+		}
+		return
+	}
+
+	proxyState := &proxyStateOnDisk{
+		Version:         Version,
```

Comment:
This is the storing the protocol version. We need a separate version to describe the state file version as that might need to change more frequently.

Reply:
Ok. I'll rework this part.